### PR TITLE
add task solution

### PR DESCRIPTION
**Pull request by maksym-nemera**
[DEMO LINK](https://maksym-nemera.github.io/react_add-todo-form/)

Have the same failed tests in [react_static-list-of-todos](https://github.com/mate-academy/react_static-list-of-todos). 
All page tests passed, but TodoInfo - failed

https://mateacademy.slack.com/archives/C050SHYFFR9/p1685359229576659

### CODE REVIEW

File: src/App.tsx [hunk 1]
```diff
@@ -1,61 +1,142 @@
...
+export const todos: Todo[] = todosFromServer.map(todo => ({
+  ...todo,
+  user: getUser(todo.userId),
+}));
 
 export const App = () => {
+  const [newTodos, setTodos] = useState<Todo[]>(todos);
+  const [title, setTitle] = useState('');
+  const [userName, setUserName] = useState('');
+  const [titleErrorMessage, setTitleError] = useState(true);
```
Comment (silchyn) [inline src/App.tsx:31]:
```suggestion
  const [isTitleValid, setIsTitleValid] = useState(true);
```

File: src/components/TodoInfo/TodoInfo.tsx [hunk 2]
```diff
@@ -1 +1,25 @@
...
+export const TodoInfo = ({ info }: Props) => {
+  const {
+    id,
+    completed,
+    title,
+    user,
+  } = info;
+
+  return (
+    <article data-id={id} className={`TodoInfo ${completed ? 'TodoInfo--completed' : ''}`}>
```
Comment (silchyn) [inline src/components/TodoInfo/TodoInfo.tsx:17]: use `cn`

**Review by silchyn — CHANGES_REQUESTED** [review]
Check behavior when the title consists of spaces

**Review by silchyn — CHANGES_REQUESTED** [review]
> Check behavior when the title consists of spaces

**Comment by NickMitchell61** [comment]
@silchyn 

if I didn't fix the issue, could you explain the issues you see a little bit more?

**Review by silchyn — CHANGES_REQUESTED** [review]
Sure, basically, it is possible to submit the form with a title consisting of only spaces, which should not be possible

File: src/App.tsx [hunk 3]
```diff
@@ -1,61 +1,142 @@
...
+  ...todo,
+  user: getUser(todo.userId),
+}));
 
 export const App = () => {
+  const [todos, setTodos] = useState<Todo[]>(preparedTodosFromServer);
+  const [title, setTitle] = useState('');
+  const [username, setUsername] = useState('');
+  const [isTitleValid, setIsTitleValid] = useState(true);
+  const [isUserValid, setIsUserValid] = useState(true);
```
Comment (silchyn) [inline src/App.tsx:32]: Is it valid initially?

File: src/App.tsx [hunk 4]
```diff
@@ -1,61 +1,142 @@
...
+
+  const handleOnSubmit = (event: React.FormEvent) => {
```
Comment (silchyn) [inline src/App.tsx:34]:
```suggestion
  const handleSubmit = (event: React.FormEvent) => {
```

File: src/App.tsx [hunk 5]
```diff
@@ -1,61 +1,142 @@
...
+  const foundUser = usersFromServer.find(user => user.id === userId);
+
+  return foundUser || null;
+}
+
+function getName(name: string): User | null {
+  const foundName = usersFromServer.find(user => user.name === name);
+
+  return foundName || null;
+}
```
Comment (silchyn) [inline src/App.tsx:20]: You are finding user, not name

File: src/App.tsx [hunk 6]
```diff
@@ -1,61 +1,142 @@
...
+  const [todos, setTodos] = useState<Todo[]>(preparedTodosFromServer);
+  const [title, setTitle] = useState('');
+  const [username, setUsername] = useState('');
+  const [isTitleValid, setIsTitleValid] = useState(true);
+  const [isUserValid, setIsUserValid] = useState(true);
+
+  const handleOnSubmit = (event: React.FormEvent) => {
+    event.preventDefault();
+
+    const maxId = Math.max(...todos.map((todo) => todo.id));
```
Comment (silchyn) [inline src/App.tsx:37]:
```suggestion
    const id = Math.max(...todos.map((todo) => todo.id)) + 1;
```

**Review by silchyn — CHANGES_REQUESTED** [review]

File: src/App.tsx [hunk 7]
```diff
@@ -1,61 +1,142 @@
...
+  const [title, setTitle] = useState('');
+  const [username, setUsername] = useState('');
+  const [isTitleValid, setIsTitleValid] = useState(false);
+  const [isUserValid, setIsUserValid] = useState(false);
+
+  const handleSubmit = (event: React.FormEvent) => {
+    event.preventDefault();
+
+    const id = Math.max(...todos.map((todo) => todo.id)) + 1;
+    const newUser = getUserByName(username);
```
Comment (silchyn) [inline src/App.tsx:38]: Is it a new user?

File: src/App.tsx [hunk 8]
```diff
@@ -1,61 +1,142 @@
...
+import { Todo } from './types/Todo';
+
+function getUserById(userId: number): User | null {
+  const foundUser = usersFromServer.find(user => user.id === userId);
+
+  return foundUser || null;
+}
+
+function getUserByName(name: string): User | null {
+  const foundName = usersFromServer.find(user => user.name === name);
```
Comment (silchyn) [inline src/App.tsx:17]: Should it be `foundName`?

**Review by silchyn — APPROVED** [review]

**Pull request closed without merging**